### PR TITLE
Conditional imports

### LINKED ISSUE #61: Conditional imports
**How to categorize this issue?**

/area landscaper
/kind enhancement
/priority normal

**What would you like to be added**:
The landscaper needs 'conditional imports' - imports that are required, if a certain other optional import is given, but are not evaluated if that other import is not specified.

**Why is this needed**:
For example the etcd backup: If activated, the landscaper needs iaas credentials and configuration for the backup blobstore. If not activated, this information is neither needed nor used. There are several more cases where some imports are only needed if another import signalizes that a specific option is activated (or similar cases).

### PR DESCRIPTION
**Pull request by Diaphteiros**
**How to categorize this PR?**

/area operations
/kind enhancement
/priority 3

**What this PR does / why we need it**:
This PR makes conditional imports work as expected.

**Which issue(s) this PR fixes**:
Fixes #61 

**Release note**:

```feature user
It is now possible to define 'conditional imports' in blueprints. Conditional imports are imports which are only checked for if another import is satisfied. See the [documentation](docs/usage/ConditionalImports.md) for further information.
```

### CODE REVIEW

File: pkg/landscaper/installations/installation.go [hunk 1]
```diff
@@ -48,14 +48,27 @@ func (i *Installation) ImportStatus() *ImportStatus {
 
 // GetImportDefinition return the import for a given key
 func (i *Installation) GetImportDefinition(key string) (lsv1alpha1.ImportDefinition, error) {
-	for _, def := range i.Blueprint.Info.Imports {
+	imports := getFlattenedImports(i.Blueprint.Info.Imports)
```
Comment (Diaphteiros) [inline pkg/landscaper/installations/installation.go:51]: @schrodit I'd like to have your opinion on this:
This check is called to see whether the parent satisfies all imports of the subinstallation. Before my changes, this looped over `i.Blueprint.Info.Imports` (over all direct imports of the parent) and checked whether the referenced import exists. With conditional imports, this is not enough anymore.

What I changed: It now loops over a flattened import list, which contains not only the direct import definitions of the parent, but also the conditional ones - independently of whether the condition is fulfilled or not.

The potential problem I see: This doesn't check whether the condition for conditional imports is fulfilled. So, in theory, an import definition could be found which doesn't actually exist, as it depends on another import which is not fulfilled. In practise, this shouldn't happen, as subinstallation imports that reference conditional parent imports are dynamically removed from the subinstallation template, if the condition which would provide the parent import is not fulfilled.

However, this still feels pretty 'dirty' and I'm not sure that I didn't miss a case in which this assumption breaks.
Should we build it in a way that the flattened list only contains conditional import definitions, if the corresponding condition is fulfilled? If so, how do we get the information which conditional imports are fulfilled at this point (`i.Imports` is `nil`)?
What do you think?

File: docs/usage/ConditionalImports.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
+# Conditional Imports
```
Comment (schrodit) [inline docs/usage/ConditionalImports.md:1]: can you link that guide also in the `Bllueprint.md`

File: pkg/landscaper/installations/imports/conditional_imports_test.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+			state *envtest.State
+		)
+
+		instRef = kutil.ObjectKey("conditional-import-inst", "test8")
+		cmRef = lsv1alpha1.ObjectReference{
+			Name:      "inst-import",
+			Namespace: instRef.Namespace,
+		}
+
+		fakeClient, state, err = envtest.NewFakeClientFromPath("./testdata/state")
```
Comment (schrodit) [inline pkg/landscaper/installations/imports/conditional_imports_test.go:55]: if the state is not needed you could also write
```suggestion
		fakeClient, _, err = envtest.NewFakeClientFromPath("./testdata/state")
```

File: pkg/landscaper/installations/imports/constructor.go [hunk 4]
```diff
@@ -67,10 +80,20 @@ func (c *Constructor) Construct(ctx context.Context, inst *installations.Install
...
 					continue // don't throw an error if the import is not required
 				}
-				return installations.NewImportNotFoundErrorf(nil, "no import for %s exists", def.Name)
+				return nil, installations.NewImportNotFoundErrorf(nil, "no import for %s exists", def.Name)
 			}
 			if err := c.JSONSchemaValidator().ValidateGoStruct(def.Schema.RawMessage, imports[def.Name]); err != nil {
-				return installations.NewErrorf(installations.SchemaValidationFailed, err, "%s: imported datatype does not have the expected schema", defPath.String())
+				return imports, installations.NewErrorf(installations.SchemaValidationFailed, err, "%s: imported datatype does not have the expected schema", defPath.String())
+			}
+			if def.ConditionalImports != nil && len(def.ConditionalImports) > 0 {
```
Comment (schrodit) [inline pkg/landscaper/installations/imports/constructor.go:88]: in go slices that are nil do also have a length of `0`
```suggestion
			if len(def.ConditionalImports) > 0 {
```

File: pkg/landscaper/installations/operation.go [hunk 5]
```diff
@@ -465,6 +469,12 @@ func (o *Operation) CreateOrUpdateImports(ctx context.Context) error {
 			return fmt.Errorf("import %s not defined", importDef.Name)
 		}
 
+		if importDef.ConditionalImports != nil && len(importDef.ConditionalImports) > 0 {
```
Comment (schrodit) [inline pkg/landscaper/installations/operation.go:472]: same as above
```suggestion
		if len(importDef.ConditionalImports) > 0 {
```

File: pkg/landscaper/installations/subinstallations/helper.go [hunk 6]
```diff
@@ -130,6 +126,27 @@ func (o *Operation) ValidateSubinstallations(installationTmpl []*lsv1alpha1.Inst
 	return nil
 }
 
+func (o *Operation) buildCoreImports(importList lsv1alpha1.ImportDefinitionList) ([]core.ImportDefinition, error) {
```
Comment (schrodit) [inline pkg/landscaper/installations/subinstallations/helper.go:129]: can you a dd a docstring explaining what the function is doing

**Review by schrodit — CHANGES_REQUESTED** [review]

**Review by schrodit — APPROVED** [review]
/lgtm

**Pull request merged**